google/cloud/pubsublite/internal/partition_publisher.h

```diff
@@ -0,0 +1,322 @@
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (shutdown_) return make_ready_future();
+      shutdown_ = true;
+      DestroyCancelToken();
+      root_future = root_future.then(ChainFuture(resumable_stream_->Shutdown()))
+                        .then([this](future<void>) {
+                          SatisfyOutstandingMessages(Status());
+                        });
+    }
+    root.set_value();
+    return root_future;
+  }
+
+ private:
+  using UnderlyingStream =
+      std::unique_ptr<AsyncStreamingReadWriteRpc<MessagePublishRequest,
+                                                 MessagePublishResponse>>;
+
+  struct MessageWithFuture {
+    PubSubMessage message;
+    promise<StatusOr<Cursor>> message_promise;
+  };
+
+  using Batch = deque<MessageWithFuture>;
+
+  static void FlattenAndMoveBatch(
+      deque<Batch>& batches,
+      deque<MessageWithFuture>& unbatched_messages_with_futures) {
+    for (auto& batch : batches) {
+      for (auto& message_with_future : batch) {
+        unbatched_messages_with_futures.emplace_back(
+            MessageWithFuture{std::move(message_with_future.message),
+                              std::move(message_with_future.message_promise)});
+      }
+    }
+    batches.clear();
+  }
+
+  // TODO(18suresha) make batching cleaner/less redundant across other functions
+  void Rebatch(bool process_in_flight) {
+    std::lock_guard<std::mutex> g{mu_};
+    deque<MessageWithFuture> unbatched_messages_with_futures;
+    if (process_in_flight) {
+      FlattenAndMoveBatch(in_flight_batches_, unbatched_messages_with_futures);
+    }
+    FlattenAndMoveBatch(batches_, unbatched_messages_with_futures);
+
+    for (auto& message_with_future : unbatched_messages_with_futures_) {
+      unbatched_messages_with_futures.emplace_back(
+          MessageWithFuture{std::move(message_with_future.message),
+                            std::move(message_with_future.message_promise)});
+    }
+    unbatched_messages_with_futures_.clear();
+
+    // rebatch
+    Batch current_batch;
+    size_t current_byte_size = 0;
+    size_t current_messages = 0;
+    for (auto& message_with_future : unbatched_messages_with_futures) {
+      size_t message_size = message_with_future.message.ByteSizeLong();
+      if (current_messages + 1 > batching_settings_.GetMessageLimit() ||
+          current_byte_size + message_size >
+              batching_settings_.GetByteLimit()) {
+        if (!current_batch.empty()) {
+          batches_.push_back(std::move(current_batch));
+          // clear because current_batch left in 'unspecified state' after move
+          current_batch.clear();
+          current_byte_size = 0;
+          current_messages = 0;
+        }
+      }
+      current_batch.push_back(std::move(message_with_future));
+      current_byte_size += message_size;
+      ++current_messages;
+    }
+    if (!current_batch.empty()) batches_.push_back(std::move(current_batch));
+  }
```

Comment:
Structure this as a member function and a free function (tested separately because it is subtle):

```
std::deque<MessageWithFuture> UnbatchAllLockHeld() {  // ABSL_LOCKS_REQUIRED(mu)
  std::deque<MessageWithFuture> to_return;
  for (auto& batch : in_flight_batches_) {
    for (auto& message : batch) {
      to_return.push_back(std::move(message));
    }
  }
  for (auto& batch : unsent_batches_) {
    for (auto& message : batch) {
      to_return.push_back(std::move(message));
    }
  }
  for (auto& message : unbatched_messages_with_futures_) {
    to_return.push_back(std::move(message));
  }
}

// Free function
std::deque<std::deque<MessageWithFuture>> CreateBatches(std::deque<MessageWithFuture> messages, const BatchingSettings& settings) {
  ... batching logic ...
}
```

Reply:
The second one can just be called as follows from the Flush()/alarm function:

unsent_batches_ = CreateBatches(std::move(unbatched_messages_with_futures_), settings);
unbatched_messages_with_futures_.clear();